src/key.cpp

```diff
@@ -251,13 +257,21 @@ bool CKey::SignCompact(const uint256 &hash, std::vector<unsigned char>& vchSig)
         return false;
     vchSig.resize(CPubKey::COMPACT_SIGNATURE_SIZE);
     int rec = -1;
-    secp256k1_ecdsa_recoverable_signature sig;
-    int ret = secp256k1_ecdsa_sign_recoverable(secp256k1_context_sign, &sig, hash.begin(), begin(), secp256k1_nonce_function_rfc6979, nullptr);
+    secp256k1_ecdsa_recoverable_signature rsig;
+    int ret = secp256k1_ecdsa_sign_recoverable(secp256k1_context_sign, &rsig, hash.begin(), begin(), secp256k1_nonce_function_rfc6979, nullptr);
     assert(ret);
-    ret = secp256k1_ecdsa_recoverable_signature_serialize_compact(secp256k1_context_sign, &vchSig[1], &rec, &sig);
+    ret = secp256k1_ecdsa_recoverable_signature_serialize_compact(secp256k1_context_sign, &vchSig[1], &rec, &rsig);
     assert(ret);
     assert(rec != -1);
     vchSig[0] = 27 + rec + (fCompressed ? 4 : 0);
+    // Additional verification step to prevent using a potentially corrupted signature
+    secp256k1_pubkey pk;
+    secp256k1_ecdsa_signature sig;
+    ret = secp256k1_ec_pubkey_create(secp256k1_context_sign, &pk, begin());
+    assert(ret);
+    secp256k1_ecdsa_recoverable_signature_convert(GetVerifyContext(), &sig, &rsig);
```

Comment:
Instead, you could invoke the recovery operation itself, and verify that the decoded public key matches `pk`. That would verify that the recovery id in the signature is correct as well (`secp256k1_ecdsa_recoverable_signature_convert` throws it away).

Reply:
Hi @sipa, thanks a lot for the input.
Let's see if I understand you right:  _instead_ of computing the `pk` in `secp256k1_ec_pubkey_create`, we recover the `pk` by running `secp256k1_ecdsa_recover` (which gives us the additional checks that you mention).
The `secp256k1_ecdsa_recoverable_signature_convert` stays because we need a normal signature to do the verification - correct?

Reply:
No, the other way around.

You still use `secp256k1_ec_pubkey_create` to compute the expected public key. But then instead of verifying it against the normal signature extracted from sig, you run recovery on the message and sig, which gives you the signing public key. Then you can compare the expected pk with the signing one.

Reply:
Perfect, thanks!